rfcs/2022-02-11-responsive-and-adaptive-layout-strategy.md

```diff
@@ -0,0 +1,394 @@
+
+## Approaches taken by well-known Design Systems
+
+| Name                                                                                                    | Responsiveness handled internally/flexibility to consumers?                                                                          |
+| ------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------ |
+| [Paste - Twilio](https://paste.twilio.design/primitives/text/#responsive-text)                          | Flexible with `[]` props accepting values for breakpoints on various components. Uses styled system under the hood                   |
+| [Primer - GitHub](https://primer.style/react/core-concepts#responsive-props)                            | Flexible with `[]` props accepting props accepting values for breakpoints on various components. Uses styled system under the hood   |
+| [Arcade](https://arcade.design/content/docs/tokens/typography)                                          | Handles internally, not customisable on the consumer end(nice strategy for less overhead)                                            |
+| [React Spectrum - Adobe](https://react-spectrum.adobe.com/react-spectrum/layout.html#responsive-layout) | Flexible to define breakpoints for the entire application and flexible to define behavior for those breakpoints per components basis |
+
+## What would work for us?
+
+As you saw there are pros and cons to each of the approach so we can't abandon either of them but we need to strike a strategy where each of these approaches will work best.
+
+After doing some research and aligning it with our uses cases at Razorpay, I propose following things at a broader level.
+
+### 1. Components that can handle Responsiveness Internally
+
+- Most of the components that don't deal with layouts shall handle responsiveness internally. For eg: Button, Modal, Avatar, Select, etc.
+
+- The properties that affects responsiveness are:
+
+  - `width`
+  - `height`
+  - `font-size`
+  - `padding`
+  - `margin`
+  - `position`
+
+- While a lot of [well known design systems gives flexibility](#approaches-taken-by-well-known-design-systems) on every component for above properties to accept responsive values, I can't think of a use case where we'll have to provide similar functionality. Majority of the cases that I could think of wouldn't be any different even if the consumers define the values for them explicitly. So seems like we'll be adding unnecessary learning curve to the component's API.
```

Comment:
This would need strict design alignment which I'm doubtful of. For example on website there are usecases for different font sizes based on breakpoints. Similarly for other properties too.

Reply:
yes so the idea here is that for typography and a few other components we'll handle it based on the defined screen sizes(breakpoints) and internally toggle based on the screen size that the app will be running on. So the consumers don't have to worry about these things everytime. If you read below I'm proposing that for few components we'll handle internally but for few like layout things we'll give the flexibility to the consumers